hail/src/main/scala/is/hail/expr/types/physical/PContainer.scala

```diff
@@ -317,22 +317,38 @@ abstract class PContainer extends PIterable {
       return Code._empty
     }
 
-    val i = mb.newLocal[Long]
+    val i = mb.newLocal[Int]
+    val mod = mb.newLocal[Int]
     Code(
-      i := PContainer.nMissingBytes(sourceType.loadLength(sourceOffset)).toL,
-      Code.whileLoop(i > 0L,
-        (i >= 8L).mux(
+      mod := (sourceType.loadLength(sourceOffset) % const(8)) - const(1),
+      (mod >= const(0)).mux(
+        Code(
+          i := PContainer.nMissingBytes(sourceType.loadLength(sourceOffset)) - const(1),
+          Code.whileLoop(mod >= const(0),
+            Code(
+              sourceType
+                .isElementMissing(sourceOffset, i * const(8) + mod)
+                .cne(const(false))
+                .orEmpty(onFail),
+              mod := mod - const(1)
+            )
+          )
+        ),
+        i := PContainer.nMissingBytes(sourceType.loadLength(sourceOffset))
+      ),
+      Code.whileLoop(i > const(0),
+        (i >= const(8)).mux(
           Code(
-            i := i - 8L,
+            i := i - const(8),
             Region
-              .loadLong(sourceOffset + sourceType.lengthHeaderBytes + i)
-              .cne(const(0.toByte))
+              .loadLong(sourceOffset + sourceType.lengthHeaderBytes + i.toL)
```

Comment:
err, this I should probably stop doing, although allocation can be many orders of magnitude slower than addition, so 🤷‍♂ should test

Reply:
I actually don't think you can load longs here. This is not necessarily an aligned read.

Reply:
What do you mean by aligned? I check that we have at least 8 elements before loading the long

Reply:
I mean aligned in the sense of "falls on an 8-byte memory boundary". Most systems *can* handle unaligned accesses, but we shouldn't do potentially unsafe things.

Reply:
I don't fully understand this. So if I check the top bits, and move down one byte, then check that I have at least 8 bytes left, how can I read something unexpected?

Reply:
Per our discussion last week, an array of, say, integers, is 4-byte aligned. This means that the missing bits could fall in between 8 byte boundaries (e.g. memory address 0xXXXXXXXXXXXX4). We assume that loadLong is called with a memory address divisible by 8, so that it is an aligned read.

Reply:
I still don't understand I think. I allocated nBytes aligned to an 8 byte boundary. I then set 8 of those bytes to 0's or 1's. I then read 8 bytes using readLong. Finally, if I allocated bytes, those are a natural alignment, so no padding. So I don't understand how reading 8 bytes at a time could have a bit that spans the boundary.

Besides this, would it matter as long as I'm reading with nMissingBytes? Lets say I did span the boundary, this would still be ok, unless I'm reading memory that I didn't allocate, which isn't the case since I take care with the top bits.

Reply:
This is the possibly unaligned read:
https://github.com/hail-is/hail/blob/04422ca5934aa4b5b24e10794a512811c638e67b/hail/src/main/scala/is/hail/expr/types/physical/PContainer.scala#L327-L328

You didn't allocate `sourceOffset`, someone else did.

Reply:
f someone allocated a PContainer, but didn't follow its intrinsics (use .allocate), wouldn't that be a bug?

Reply:
yes, but the semantics of PContainer.allocate don't guarantee that the missing bits are 8-byte aligned.

Reply:
I see, so you're suggesting an int because of `Region.storeInt(aoff, length)` in initialize?

Since after that int, we allocate individual bytes, would it then be sufficient to read 1 int (if possible), and then switch to 8 byte values (since a memory address xxx4 would increment by 1 for each byte)?